msgpack/src/main/scala/fs2/data/msgpack/low/internal/ItemSerializer.scala

```diff
@@ -0,0 +1,247 @@
+  class MalformedUintError extends MalformedItemError
+
+  /** Checks whether integer `x` fits in `n` bytes. */
+  @inline
+  private def fitsIn(x: Int, n: Long): Boolean =
+    java.lang.Integer.compareUnsigned(x, (Math.pow(2, n.toDouble).toLong - 1).toInt) <= 0
+
+  private case class SerializationContext[F[_]](out: Out[F],
+                                                chunk: Chunk[MsgpackItem],
+                                                idx: Int,
+                                                rest: Stream[F, MsgpackItem])
+
+  /** Buffers [[Chunk]] into 4KiB segments before calling [[Pull.output]].
+    *
+    * @param contents buffered [[Chunk]]
+    */
+  private class Out[F[_]](contents: Chunk[Byte]) {
+    private val limit = 4096
+
+    /** Pushes `bv` into the buffer and emits the buffer if it reaches the limit.
+     */
+    @inline
+    def push(bv: ByteVector): Pull[F, Byte, Out[F]] =
+      if (contents.size >= limit)
+        Pull.output(contents).as(new Out(Chunk.byteVector(bv)))
+      else
+        Pull.done.as(new Out(contents ++ Chunk.byteVector(bv)))
+
+    /** Splices `bv` into segments and pushes them into the buffer while emitting the buffer at the same time so
+      * that it never exceeds the limit during the operation.
+      *
+      * Use this instead of [[Out.push]] when `bv` may significantly exceed 4KiB.
+      */
+    def pushBuffered(bv: ByteVector): Pull[F, Byte, Out[F]] = {
+      @inline
+      def go(chunk: Chunk[Byte], rest: ByteVector): Pull[F, Byte, Out[F]] =
+        if (rest.isEmpty)
+          Pull.done.as(new Out(chunk))
+        else
+          Pull.output(chunk) >> go(Chunk.byteVector(rest.take(limit.toLong)), rest.drop(limit.toLong))
+
+      if (bv.isEmpty)
+        this.push(bv)
+      else if (contents.size >= limit)
+        Pull.output(contents) >> go(Chunk.byteVector(bv.take(limit.toLong)), bv.drop(limit.toLong))
+      else
+        go(contents ++ Chunk.byteVector(bv.take(limit.toLong - contents.size)), bv.drop(limit.toLong - contents.size))
+    }
+
+    /** Outputs the whole buffer. */
+    @inline
+    def flush = Pull.output(contents)
+  }
+
+  @inline
+  private def step[F[_]: RaiseThrowable](o: Out[F], item: MsgpackItem): Pull[F, Byte, Out[F]] = item match {
+    case MsgpackItem.UnsignedInt(bytes) =>
+      val bs = bytes.dropWhile(_ == 0)
+      if (bs.size <= 1)
+        o.push(ByteVector(Headers.Uint8) ++ bs.padLeft(1))
+      else if (bs.size <= 2)
+        o.push(ByteVector(Headers.Uint16) ++ bs.padLeft(2))
+      else if (bs.size <= 4)
+        o.push(ByteVector(Headers.Uint32) ++ bs.padLeft(4))
+      else if (bs.size <= 8)
+        o.push(ByteVector(Headers.Uint64) ++ bs.padLeft(8))
+      else
+        Pull.raiseError(new MalformedUintError)
+
+    case MsgpackItem.SignedInt(bytes) =>
+      val bs = bytes.dropWhile(_ == 0)
+      if (bs.size <= 1)
+        // positive fixint or negative fixint
+        if ((bs & hex"7f") == bs || (bs & hex"c0") == hex"c0")
```

Comment:
We could extract the constants (e.g. `hex"7f"`) into object fields to avoid reallocating a `ByteVector` each time.